File: src/main/java/AoC2017_10.java
import static java.util.stream.Collectors.toList;

import com.github.pareronia.aoc.knothash.KnotHash;
import com.github.pareronia.aoc.knothash.KnotHash.State;
import com.github.pareronia.aoc.solution.Sample;
import com.github.pareronia.aoc.solution.Samples;
import com.github.pareronia.aoc.solution.SolutionBase;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

@SuppressWarnings({"PMD.ClassNamingConventions", "PMD.NoPackage"})
public final class AoC2017_10 extends SolutionBase<String, Integer, String> {

    private AoC2017_10(final boolean debug) {
        super(debug);
    }

    public static AoC2017_10 create() {
        return new AoC2017_10(false);
    }

    public static AoC2017_10 createDebug() {
        return new AoC2017_10(true);
    }

    @Override
    protected String parseInput(final List<String> inputs) {
        return inputs.getFirst();
    }

    private Integer solve1(final String input, final List<Integer> elements) {
        final List<Integer> lengths =
                Arrays.stream(input.split(",")).map(Integer::valueOf).collect(toList());
        final State state = new State(elements, lengths, 0, 0);
        final State ans = KnotHash.round(state);
        return ans.elements().get(0) * ans.elements().get(1);
    }

    @Override
    public Integer solvePart1(final String input) {
        return solve1(input, new ArrayList<>(KnotHash.SEED));
    }

    @Override
    public String solvePart2(final String input) {
        return KnotHash.hexString(input);
    }

    @Override
	protected void samples() {
    	final AoC2017_10 test = createDebug();
    	final String input = test.parseInput(List.of("3,4,1,5"));
    	assert test.solve1(input, new ArrayList<>(List.of(0, 1, 2, 3, 4))) == 12;
	}

	@Samples({
        @Sample(
                method = "part2",
                input = "AoC 2017",
                expected = "33efeb34ea91902bb2f59c9920caa6cd"),
        @Sample(method = "part2", input = "1,2,3", expected = "3efbe78a8d82f29979031a4aa0b16a9d"),
        @Sample(method = "part2", input = "1,2,4", expected = "63960835bcdc130f0b66d7ff4f6a5a8e"),
    })
    public static void main(final String[] args) throws Exception {
        create().run();
    }
}

File: src/main/python/AoC2017_10.py
#! /usr/bin/env python3
#
# Advent of Code 2017 Day 10
#

from aoc import my_aocd
from aoc.knothash import KnotHash
import aocd


def _solve_1(inputs: tuple[str], elements: list[int]) -> int:
    lengths = [int(_) for _ in inputs[0].split(",")]
    ans = KnotHash.round(KnotHash.State(elements, lengths, 0, 0))
    return ans.elements[0] * ans.elements[1]


def part_1(inputs: tuple[str]) -> int:
    return _solve_1(inputs, [_ for _ in KnotHash.SEED])


def part_2(inputs: tuple[str]) -> int:
    return KnotHash.hex_string(inputs[0])


TEST1 = """3,4,1,5""".splitlines()
TEST2 = """AoC 2017""".splitlines()
TEST3 = """1,2,3""".splitlines()
TEST4 = """1,2,4""".splitlines()


def main() -> None:
    puzzle = aocd.models.Puzzle(2017, 10)
    my_aocd.print_header(puzzle.year, puzzle.day)

    assert _solve_1(TEST1, [0, 1, 2, 3, 4]) == 12
    assert part_2(TEST2) == "33efeb34ea91902bb2f59c9920caa6cd"
    assert part_2(TEST3) == "3efbe78a8d82f29979031a4aa0b16a9d"
    assert part_2(TEST4) == "63960835bcdc130f0b66d7ff4f6a5a8e"

    inputs = my_aocd.get_input(2017, 10, 1)
    result1 = part_1(inputs)
    print(f"Part 1: {result1}")
    result2 = part_2(inputs)
    print(f"Part 2: {result2}")
    my_aocd.check_results(puzzle, result1, result2)


if __name__ == "__main__":
    main()
